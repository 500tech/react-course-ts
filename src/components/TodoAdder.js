import React, { useState, useRef, useEffect } from 'react';
import { NOOP } from '../utils';
import { Button, Input } from './atoms';

export function TodoAdder({ onAddTodo = NOOP }) {
  const [draft, setDraft] = useState('');
  const inputRef = useRef();
  useEffect(() => {
    const tid = setTimeout(() => console.log('No input for a while...'), 3000);
    return () => clearTimeout(tid);
  }, [draft]);
  useEffect(() => {
    const actualInputNode = inputRef.current;
    actualInputNode.focus();
  }, []);
  return (
    <>
      <Input
        ref={inputRef}
        value={draft}
        onChange={event => setDraft(event.target.value)}
      />
      <Button
        onClick={() => {
          onAddTodo(draft);
          setDraft('');
        }}
        disabled={!draft}
      >
        Add Todo
      </Button>
    </>
  );
}
