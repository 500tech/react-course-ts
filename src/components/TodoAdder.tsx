import React, { useState, useRef, useEffect, useCallback } from "react";
import styled, { css } from "styled-components";

interface IProps {
  onAddTodo: (text: string) => void;
}

function useFormState<T>(initialState: string) {
  const [draft, setDraft] = useState(initialState);
  const setOnChange = (e: React.ChangeEvent<HTMLInputElement>) =>
    setDraft(e.target.value);

  const controlled = { value: draft, onChange: setOnChange };
  return [draft, controlled, setDraft] as [
    string,
    typeof controlled,
    typeof setDraft
  ];
}

function useAutofocus() {
  const inputRef = useRef<HTMLInputElement>(null);
  useEffect(() => {
    inputRef.current!.focus();
  }, []);
  return { ref: inputRef };
}

const disabledStyle = css`
  border: none;
  box-shadow: none;
`;

const enabledStyle = css`
  background-color: pink;
`;

const Button = styled.button`
  margin: 0 10px;
  ${props => (props.disabled ? disabledStyle : enabledStyle)}
`;

const Input = styled.input`
  &:focus {
    outline-color: ${props => props.theme.colors.primary};
  }
`;

export const TodoAdder2: React.FC<IProps> = ({ onAddTodo }) => {
  const [draft, controlled, setDraft] = useFormState("");
  const autofocus = useAutofocus();

  const submit = useCallback(
    (e?: React.FormEvent) => {
      e && e.preventDefault();
      if (draft) {
        onAddTodo(draft);
        setDraft("");
      }
    },
    [draft, onAddTodo, setDraft]
  );

  useEffect(() => {
    const tid = setTimeout(submit, 1500);
    return () => clearTimeout(tid);
  }, [draft, submit]);

  return (
    <form onSubmit={submit}>
      <Input
        placeholder="Thing to do..."
        type="text"
        {...autofocus}
        {...controlled}
      />
      <Button type="submit" disabled={!draft}>
        Add
      </Button>
    </form>
  );
};
