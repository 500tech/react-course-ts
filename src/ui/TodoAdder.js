import React, { Component, createRef } from 'react';
import { RudeButton } from './common';

export class TodoAdder extends Component {
  state = {
    text: '',
  };

  inputRef = createRef();

  componentDidMount() {
    this.inputRef.current.focus();
    this.setAutosubmit();
  }

  componentDidUpdate(_prevProps, prevState) {
    if (prevState.text !== this.state.text) {
      this.setAutosubmit();
    }
  }

  componentWillUnmount() {
    clearTimeout(this.timerId);
  }

  setAutosubmit() {
    clearTimeout(this.timerId);
    this.timerId = setTimeout(this.submit, 3000);
  }

  get canSubmit() {
    const { text } = this.state;
    return text.length > 0 && !text.match(/password/i);
  }

  updateText = e => {
    const { value } = e.target;
    if (value.match(/clear/i)) {
      return this.setState({ text: '' });
    }
    this.setState({ text: value });
  };

  submit = e => {
    e && e.preventDefault();
    const { text } = this.state;
    if (this.canSubmit) {
      this.props.onAdd(text);
      this.setState({ text: '' });
    }
  };

  render() {
    const { text } = this.state;
    return (
      <form onSubmit={this.submit}>
        <input
          ref={this.inputRef}
          placeholder="Enter todo here..."
          value={text}
          onChange={this.updateText}
        />
        <RudeButton disabled={!this.canSubmit}>Add</RudeButton>
      </form>
    );
  }
}
