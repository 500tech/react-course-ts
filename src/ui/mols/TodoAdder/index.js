import React, { Component, createRef } from 'react';

export class TodoAdder extends Component {
  constructor(props, ...args) {
    super(props, ...args);
    this.state = {
      text: props.initialText || '',
    };
  }

  inputRef = createRef();

  updateText = e => {
    this.setState({ text: e.target.value });
  };

  submit = e => {
    e && e.preventDefault();
    const { onAdd } = this.props;
    const { text } = this.state;
    if (text) {
      onAdd && onAdd(text);
      this.setState({ text: '' });
    }
  };

  componentDidMount() {
    this.inputRef.current.focus();
  }

  render() {
    return (
      <form onSubmit={this.submit}>
        <input
          value={this.state.text}
          onChange={this.updateText}
          ref={this.inputRef}
        />
        <button disabled={!this.state.text}>Add</button>
      </form>
    );
  }
}
