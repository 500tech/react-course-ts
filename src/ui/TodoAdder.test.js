import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { TodoAdder } from './TodoAdder';

describe('TodoAdder', () => {
  it('render', () => {
    render(<TodoAdder />);
  });

  it('is the same as before', () => {
    expect(render(<TodoAdder />).asFragment()).toMatchSnapshot();
  });

  it('has auto focus', () => {
    const { queryByPlaceholderText } = render(<TodoAdder />);
    const input = queryByPlaceholderText(/enter todo/i);
    expect(input).not.toBeNull();
    expect(input).toHaveFocus();
  });

  it('changes input text on type', () => {
    const { queryByPlaceholderText } = render(<TodoAdder />);
    const input = queryByPlaceholderText(/enter todo/i);
    fireEvent.change(input, {
      target: {
        value: 'hello',
      },
    });
    expect(input).toHaveValue('hello');
    fireEvent.change(input, {
      target: {
        value: 'hello clear',
      },
    });
    expect(input).toHaveValue('');
  });

  it('only enables and shows the button when applicable', () => {
    const { queryByPlaceholderText, getByText } = render(<TodoAdder />);
    const input = queryByPlaceholderText(/enter todo/i);
    const btn = getByText(/add/i);
    expect(btn).not.toBeVisible();
    fireEvent.change(input, {
      target: {
        value: 'hello',
      },
    });
    expect(btn).toBeVisible();
    fireEvent.change(input, {
      target: {
        value: 'password',
      },
    });
    expect(btn).not.toBeVisible();
  });

  it('calls on add when it should', () => {
    const onAdd = jest.fn();
    const { queryByPlaceholderText } = render(<TodoAdder onAdd={onAdd} />);
    const input = queryByPlaceholderText(/enter todo/i);
    fireEvent.submit(input);
    expect(onAdd).not.toHaveBeenCalled();
    fireEvent.change(input, {
      target: {
        value: 'hello',
      },
    });
    fireEvent.submit(input);
    expect(onAdd).toHaveBeenCalledWith('hello');
    expect(input).toHaveValue('');
  });
});
