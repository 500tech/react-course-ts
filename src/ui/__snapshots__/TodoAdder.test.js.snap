// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoAdder is the same as before 1`] = `
<DocumentFragment>
  <form>
    <input
      placeholder="Enter todo here..."
      value=""
    />
    <button
      class="sc-bdVaJa etGpg"
      disabled=""
    >
      Add
    </button>
  </form>
</DocumentFragment>
`;
